File: sdk/nodejs1/sdk/src/zrok/access.ts
import {BackendMode} from "./share";
import {Root} from "./environment";
import {instanceOfCreateFrontend201Response, ShareApi} from "../api";

export class AccessRequest {
    shareToken: string;

    constructor(shareToken: string) {
        this.shareToken = shareToken;
    }
}

export class Access {
    frontendToken: string;
    shareToken: string;
    backendMode: BackendMode;

    constructor(frontendToken: string, shareToken: string, backendMode: BackendMode) {
        this.frontendToken = frontendToken;
        this.shareToken = shareToken;
        this.backendMode = backendMode;
    }
}

export const createAccess = async (root: Root, request: AccessRequest): Promise<Access> => {
    if(!root.isEnabled()) {
        throw new Error("environment is not enabled; enable with 'zrok enable' first!");
    }

    let acc = await new ShareApi(root.apiConfiguration()).access({body: {
            envZId: root.environment?.zId,
            shareToken: request.shareToken
        }});

    return new Access(acc.frontendToken!, request.shareToken, acc.backendMode!);
}

export const deleteAccess = async (root: Root, acc: Access): Promise<any> => {
    if(!root.isEnabled()) {
        throw new Error("environment is not enabled; enable with 'zrok enable' first!");
    }
    return new ShareApi(root.apiConfiguration()).unaccess({body: {
            envZId: root.environment?.zId,
            shareToken: acc.shareToken,
            frontendToken: acc.frontendToken
        }});
}

File: sdk/golang/sdk/access.go
package sdk

import (
	httptransport "github.com/go-openapi/runtime/client"
	"github.com/openziti/zrok/environment/env_core"
	"github.com/openziti/zrok/rest_client_zrok/share"
	"github.com/pkg/errors"
)

func CreateAccess(root env_core.Root, request *AccessRequest) (*Access, error) {
	if !root.IsEnabled() {
		return nil, errors.New("environment is not enabled; enable with 'zrok enable' first!")
	}

	out := share.NewAccessParams()
	out.Body.ShareToken = request.ShareToken
	out.Body.EnvZID = root.Environment().ZitiIdentity

	zrok, err := root.Client()
	if err != nil {
		return nil, errors.Wrap(err, "error getting zrok client")
	}
	auth := httptransport.APIKeyAuth("X-TOKEN", "header", root.Environment().AccountToken)

	in, err := zrok.Share.Access(out, auth)
	if err != nil {
		return nil, errors.Wrap(err, "unable to create access")
	}

	return &Access{Token: in.Payload.FrontendToken, ShareToken: request.ShareToken, BackendMode: BackendMode(in.Payload.BackendMode)}, nil
}

func DeleteAccess(root env_core.Root, acc *Access) error {
	out := share.NewUnaccessParams()
	out.Body.FrontendToken = acc.Token
	out.Body.ShareToken = acc.ShareToken
	out.Body.EnvZID = root.Environment().ZitiIdentity

	zrok, err := root.Client()
	if err != nil {
		return errors.Wrap(err, "error getting zrok client")
	}
	auth := httptransport.APIKeyAuth("X-TOKEN", "header", root.Environment().AccountToken)

	_, err = zrok.Share.Unaccess(out, auth)
	if err != nil {
		return errors.Wrap(err, "error deleting access")
	}

	return nil
}
